sockeye/data_io.py

```diff
@@ -479,24 +534,76 @@ def _assign_to_buckets(self, source_sentences, target_sentences):
         logger.info("fill up mode: %s", self.fill_up)
         logger.info("")
 
+    def _populate_bucket_batch_sizes(self):
+        """
+        Compute bucket-specific batch sizes (sentences, average_words) and default bucket batch
+        size.
+
+        If sentence-based batching: number of sentences is the same for each batch, determines the
+        number of words.
+
+        If word-based batching: number of sentences for each batch is set to the multiple of number
+        of devices that produces the number of words closest to the target batch size.  Average
+        target sentence length (non-padding symbols) is used for word number calculations.
+
+        Default bucket batch size is the number of sentences such that shape
+        (default_bucket_batch_size * default_bucket_key) will fit the total number of elements in
+        the largest batch (batch_size * seq_len).
+
+        Sets: self.bucket_batch_sizes, self.default_bucket_batch_size
+        """
+        # Pre-defined bucket batch sizes
+        if self.bucket_batch_sizes is not None:
+            return
+        # Otherwise compute here
+        self.bucket_batch_sizes = [[] for _ in self.buckets]
+        for buck_idx, bucket_shape in enumerate(self.buckets):
+            # Target/label length with padding
+            padded_seq_len = bucket_shape[1]
+            # Average target/label length excluding padding
+            average_seq_len = self.data_label_average_len[buck_idx]
+            if self.batch_by_words:
+                check_condition(padded_seq_len <= self.batch_size, "Word batch size must cover sequence lengths for all"
+                                " buckets: (%d > %d)" % (padded_seq_len, self.batch_size))
+            # Word-based: num words determines num sentences
+            # Sentence-based: num sentences determines num words
+            if self.batch_by_words:
+                # Multiple of number of devices (int) closest to target number of words, assuming each sentence is of
+                # average length
+                batch_size_seq = self.batch_num_devices * round((self.batch_size / average_seq_len)
+                                                                / self.batch_num_devices)
+                batch_size_word = batch_size_seq * average_seq_len
+            else:
+                batch_size_seq = self.batch_size
+                batch_size_word = batch_size_seq * average_seq_len
+            self.bucket_batch_sizes[buck_idx] = (batch_size_seq, batch_size_word)
+        # Default bucket batch size: batch size of largest bucket batch by total elements, scaled for default bucket key
+        # Effectively, set default values to reserve enough space to provide a batch from any bucket
+        largest_total_batch_size = 0
+        for buck_shape, (batch_size_seq, _) in zip(self.buckets, self.bucket_batch_sizes):
```

Comment:
for sentence based batch sizes this is essentially a noop, no? could you make this explicit? something like:
```
if self.batch_by_words:
  for ...
else:
  self.default_bucket_batch_size = self.batch_size
```